Adds support to Postgres

pgeadas opened the pull request:
## Description 

Adds support for `postgres` driver:
- org.postgresql.Driver


## Caveats
Backticks are a non-standard used by MySQL. Postgres does not support them by default.

pditommaso commented on src/main/java/io/seqera/migtool/MigTool.java:
```diff
@@ -447,7 +462,12 @@ private int migrate(MigRecord entry) throws SQLException {
     protected boolean checkMigrated(MigRecord entry) {
-        String sql = "select `id`, `checksum`, `script` from " + MIGTOOL_TABLE + " where `rank` = ? and `script` = ?";
+        String sql;
+        if(dialect.isPostgres() || dialect.isTestContainersPostgres()) {
+            sql = "select id, checksum, script from " + MIGTOOL_TABLE + " where rank = ? and script = ?";
+        } else {
+            sql = "select `id`, `checksum`, `script` from " + MIGTOOL_TABLE + " where `rank` = ? and `script` = ?";
+        }
```
Why does Postgres require special handling?

pgeadas replied:
I added in the PR description.

> Backticks are a non-standard used by MySQL. Postgres does not support them by default, so the simple option was to add a conditional to check for the driver type and remove the backticks accordingly.

pgeadas committed:
Adds support to Postgres

tcrespog commented on src/main/java/io/seqera/migtool/Driver.java:
```diff
@@ -0,0 +1,29 @@
+    H2("org.h2.Driver"),
+    SQLITE("org.sqlite.JDBC"),
+    POSTGRES("org.postgresql.Driver"),
+    TCPOSTGRES("org.testcontainers.jdbc.ContainerDatabaseDriver");
```
As far as I can see, this driver initializes a container database directly behind the scenes just by using a `jdbc:tc:...` connection URL.

Given that we already take care of starting a TC container independently in tests, I think it might not be worth adding support for it. Do you think it's worth?

pgeadas replied:
It is probably not needed, I can delete it.

tcrespog commented on src/main/java/io/seqera/migtool/MigTool.java:
```diff
@@ -42,13 +44,13 @@ public class MigTool {
     static final String MIGTOOL_TABLE = "MIGTOOL_HISTORY";
 
-    static final String[] DIALECTS = {"h2", "mysql", "mariadb","sqlite"};
+    static final Dialect[] DIALECTS = {H2, MYSQL, MARIADB, SQLITE, POSTGRES, TCPOSTGRES};
```
Given that an enum is used now, this might not be needed anymore.

tcrespog commented on src/main/java/io/seqera/migtool/Helper.java:
```diff
@@ -201,21 +201,31 @@ private static void closeWithWarning(Closeable closeable) {
+        if( dialect==null ) {
+            return null;
+        }
+        switch (dialect) {
```
Instead of the `switch`, maybe we could add a `driver` property in the `Dialect` enum (with its corresponding `getDriver()`/`toDriver` getter) initializing it accordingly in the constructor of the enum.

pgeadas replied:
Yes, that can be done. I just didn't want to refactor the entire code yesterday, but I'm ok improving it a bit more :)

tcrespog commented on src/main/java/io/seqera/migtool/MigTool.java:
```diff
@@ -350,7 +360,9 @@ protected void apply() {
     protected void checkRank(MigRecord entry) {
         try(Connection conn=getConnection(); Statement stm = conn.createStatement()) {
-            ResultSet rs = stm.executeQuery("select max(`rank`) from "+MIGTOOL_TABLE);
+            String rank = dialect.isPostgres() || dialect.isTestContainersPostgres()  ? "rank" : "`rank`";
```
As an alternative to defining `column` and `\`column`\`, we could add a method such as `transformColumnsToPostgres` which removes the backticks. Ideally, the method will have a comment explaining that backticks are not supported in PostgreSQL.

pditommaso replied:
I agree this should be improved. Also, it could be considered a kind of "driver" pattern.

pgeadas replied:
Actually I think that a strategy pattern can be applied easily here, however I went for the fast fix. I will improve the solution then.

pgeadas committed:
Removed 'test containers driver'; small improvements

pgeadas committed:
Refactored MigTool to use a standard Builder pattern

pgeadas committed:
Refactored MigTool to use a query builder

pgeadas committed:
Improve documentation

tcrespog commented on src/main/java/io/seqera/migtool/Dialect.java:
```diff
@@ -0,0 +1,60 @@
+package io.seqera.migtool;
+
+import groovyjarjarantlr4.v4.runtime.misc.Nullable;
```
This `Nullable` annotation comes from a Groovy package, any chance of using other? In the end, importing `Nullable` is [not as trivial](https://stackoverflow.com/questions/76630457/what-nullable-to-use-in-java-as-of-2023-jdk21) as it might seem. Alternatively, we could just not use it.

pgeadas replied:
In the end, the annotation is just a marker so that people know that they should check for or expect a null value. Do we care from which package it comes from? If so, I can remove it, but probably it is ok ...

tcrespog replied:
Okay, I was saying that because it seems to come from a somewhat obscure Groovy package that may cause confusion, but I personally don't mind.

pgeadas replied:
I'll see if I can a more known one instead.

tcrespog commented on README.md:
````diff
@@ -25,33 +25,33 @@ If for a given file there is only either a single .patch or a single .override t
 
 ```groovy
-    MigTool tool = new MigTool();
+    MigTool tool = new MigTool.Builder()
````
We should remember that users of the library (such as Platform) will need to be updated to the new building interface once the new version is applied to them.

pgeadas replied:
I think users will notice when they upgrade! 😃 Luckily it is a very small change.

pgeadas committed:
Small fixes

Added jakarta annotations; added new constructor to be compatible with Java; changed the visibility of a method, to avoid the need of extending the base class just to use it

pgeadas commented:
Are we merging this at some point? @swampie @pditommaso @tcrespog

tcrespog commented:
> Are we merging this at some point? @swampie @pditommaso @tcrespog

It's okay for me (I approved it). Should we wait for @swampie and @pditommaso to check it? Please, Matteo & Paolo, let us know.

pditommaso commented on src/main/java/io/seqera/migtool/MigTool.java:
```diff
@@ -293,23 +276,18 @@ protected void scanMigrations() {
-                while( itr.hasNext() ) {
-                    Path it = itr.next();
+            try (DirectoryStream<Path> stream = Files.newDirectoryStream(Paths.get(path))) {
+                for (Path it : stream) {
```
WHat's the value of this change? don't see related to the support for Postgress

pditommaso commented on src/main/java/io/seqera/migtool/MigTool.java:
```diff
@@ -509,4 +487,64 @@ public Object doCall() {
+        }
+
+        public Builder withUser(String user) {
+            this.user = user;
+            return this;
+        }
+
+        public Builder withPassword(String password) {
+            this.password = password;
+            return this;
+        }
+
+        public Builder withDialect(String dialectName) {
+            this.dialectName = dialectName;
+            return this;
+        }
+
+        public Builder withLocations(String locations) {
+            this.locations = locations;
+            return this;
+        }
+
+        public Builder withClassLoader(ClassLoader loader) {
+            this.classLoader = loader;
+            return this;
+        }
+
+        public Builder withPattern(String pattern) {
+            if(pattern != null && !pattern.isEmpty())
+                this.pattern = Pattern.compile(pattern);
+            return this;
+        }
+
+        public MigTool build() {
+            this.driver = driverName != null ? Driver.fromDriverName(driverName) : Driver.fromUrl(url);
+            this.dialect = dialectName != null ? Dialect.fromDialectName(dialectName) : Dialect.fromUrl(url);
+            return new MigTool(this);
+        }
+    }
+
```
Bit useless. It was avoid on purpose to not implement the builder pattern because it's just redundant

pditommaso commented on src/main/java/io/seqera/migtool/strategy/DialectQueryBuilder.java:
```diff
@@ -0,0 +1,48 @@
+ * Since the other dialects supported allow the use of backticks, only two builders are added
+ * {@link PostgresQueryBuilder} and {@link NotPostgresQueryBuilder}.
+ **/
+public abstract class DialectQueryBuilder {
```
I found this solution overkill. It would be enough to have a driver pattern having a default implementation (a couple of methods) covering the current statement and the version for Postgres.

pditommaso replied:
That kind of things that's simpler to implement, that to describe. If we want to adhere to GoF definition, it would be a [template pattern](https://www.digitalocean.com/community/tutorials/template-method-design-pattern-in-java), but also Jdbc driver using a similar approach, this is why I had used that term. 

Compare [this](https://github.com/seqeralabs/migtool/pull/49) implementation with the one in this PR (I hadn't time to add tests)

pgeadas replied:
I considered this approach, but I don't like it because the common SQL needs to be duplicated in both implementations. Meaning that if we need to change one of the queries in one place, we need to change it in both files (Default and Postgres).

In the end, we also needed to add 3 classes here, so I don't see what we gained using this approach. :)

I am proposing a solution now that solves both issues: the one regarding being overkill (since we only have 2 implementations) and the one that implies having the sql string repeated in the subclasses.

Would this be acceptable in your opinion? I think it merges the best of both worlds @pditommaso .

pditommaso commented on src/main/java/io/seqera/migtool/Dialect.java:
```diff
@@ -0,0 +1,60 @@
+import jakarta.annotation.Nullable;
+import java.util.List;
+
+public enum Dialect {
```
Not sure what value adds `Dialect` and `Driver` as enum. it would make more sense to implement a driver pattern that can both the type and the method specialisation for the statement specific for postgres. Likely it would require a `Driver` interface, plus an abstract implementation for the common stuff, and concrete class for each support DB. 

An alternative approach could be using a delegate pattern and passing the concrete class as "argument"

pgeadas committed:
refactor: simplify how to choose the sql template

pgeadas commented on src/main/java/io/seqera/migtool/MigRecord.java:
```diff
@@ -40,6 +40,16 @@ static Language from(String fileName) {
     boolean isPatch;
     boolean isOverride;
 
+    /**
```
For example this is missing @pditommaso , otherwise it won't work from Java...

pditommaso replied:
Is this to Postgres?

pgeadas replied:
> Is this related to Postgres?

<img width="824" alt="image" src="https://github.com/user-attachments/assets/5b56dcc6-5b47-46eb-95c4-74519c3d1811">

This won't work from Java, but I guess it is not needed for Postgres.

pditommaso commented:
Sorry Pedro, I'm closing in favour of #49. While I'm sure this work correctly, I believe is not a good PR because it conflates together many different things: support for Postres, support for Java, general refactoring, unnecessary complexity, changes the MigTool public API, etc. 

Why I think this is bad: 

- make difficult if not impossible to make an effective code review, because there are too many unrelated changes;
- make it difficult to recreate the history and the rationale of the changes;
- it can introduce unexpected bugs or side-effects; 
- it's contrary to simplicity and frugality principles;
- last and even more importantly, software source code is a collaborative effort and shared knowledge. it represents a mental model for all the people that have worked with it and know how to deal with it. Making a radical refactoring - without a substantial reason - it means that all these people (me, Thomas, Veronica,  whoever) will need to spend time to learn new data structures, new behaviours, etc for no clear reason. 


I don't want to be further pedantic, but I'b be happy to discuss more if needed.

pgeadas commented:
> * make difficult if not impossible to make an effective code review, because there are too many unrelated changes;

This point I can agree with. Some changes are unrelated to Postgres and were done just to make it easier to implement that change. I can split this in different PR's if this is the issue.

Three PR's refactors with: 
- adding the enums instead of hardcoded Strings in the code
- adding the real builder pattern (as it is, we need to do things from the outside to build the MigTool every time)
- adding the real template pattern (as it is, the queries are not following any template, if the query changes we need to manually change two classes)

But this is by no means a radical refactor, this is just a cleanup. 

> it's contrary to simplicity and frugality principles; 

I don't think hardcoded strings and methods that do 3 or 4 things at the same time, can be considered simplicity.

> it means that all these people (me, Thomas, Veronica, whoever) will need to spend time to learn new data structures, new behaviours, etc for no clear reason.
> unnecessary complexity

I think an enum to replace hardcoded Strings and encapsulate behaviours that right now exist on a static helper class, is not "for no reason" and _removes_ complexity, does not increase it. 🤷🏻‍♂️ 
In fact, I assure you it would be way easier to understand for new people and even the old ones when they need to come back to it.

> last and even more importantly, software source code is a collaborative effort and shared knowledge. 

I don't think that I am the one missing this point here 👼🏻 The PR already has several approvals from other team members... 

Actually me and @tcrespog had discussed this improvements together, in a collaborative way. 

> I don't want to be further pedantic, but I'b be happy to discuss more if needed.

I could give you all the reasons in the world that it would not have made a difference, so I don't think it is worth discussing more, you are right @pditommaso 🥲
